participants/l10n_update_ts.py

```diff
@@ -226,22 +236,28 @@ def init_gitdir(self, reponame):
                     'has_downloads': False,
                     'auto_init': True
                 }
-                ghresp = requests.post("%s/user/repos" % self.gitconf["apiurl"],
-                                       auth=gitserv_auth,
-                                       headers={
-                                           'content-type': 'application/json'
-                                       },
-                                       verify=False,
-                                       data=json.dumps(payload))
+                ghresp = requests.post(
+                    "%s/user/repos" % self.gitconf["apiurl"],
+                    auth=gitserv_auth,
+                    headers={'content-type': 'application/json'},
+                    verify=False,
+                    data=json.dumps(payload)
+                )
                 ghresp.raise_for_status()
 
-            check_call(["git", "clone",
-                        self.gitconf["repourl"] % {
-                            "username": self.gitconf["username"],
-                            "reponame": reponame
-                        }],
+            repo_url = self.gitconf["repourl"] % {
+                "username": self.gitconf["username"],
+                "reponame": reponame
+            }
+            check_call(["git", "clone", repo_url, gitdir],
                        cwd=self.gitconf["basedir"])
+            current_branch = check_output(
+                ['git', 'symbolic-ref', '-q', 'HEAD'], cwd=gitdir).strip()
+            if current_branch.replace('refs/heads/', '') != branch:
+                check_call(["git", "branch", branch], cwd=gitdir)
```

Comment:
IIUC, this would create a new branch, even if remote branch exists.
If a local branch exists too, this command will throw an exception.

Reply:
agreed - if the remote is active on branchA but has branchB and we're using branchB then branchB locally will point to HEAD which is branchA remotely.
(aside: since we've just cloned I don't think a local branch can exist)
Since the logic is "if the current branch is not the desired branch then checkout the desired branch"... and given this is a fresh clone ... why don't we just do an unconditional "git checkout -b branch --track origin/branch" (ie an explicitly expanded "git checkout branch").
This of course may throw an error if there is no remote/branch so either catch that and complain or test for origin/branch and complain :)

The git reset --hard is also then redundant since we do a nice clean tracking checkout to the desired branch.

Reply:
I did it like this because the branch might exists or not in the remote, so this would work in both cases without extra checks or error handling... Except now I see that it will probably fail in last 'git reset' step if the branch does not exist in the remote. So yes, checking the remote branch existence first and checkout from that directly is probably better.